### ISSUE TITLE
Upgrade migrations did not complete when attempting to upgrade from 3.6.17 to 3.7.7

### ISSUE THREAD

**Issue by croxton** [craftcms/cms#9660]
### Description

Appears to be similar to the issue as reported here, relating to Global Sets: #8079.

I will revert and see if there is anything odd about the global sets.

```
*** applying m210613_145522_sortable_global_sets
    > add column sortOrder smallint UNSIGNED AFTER `fieldLayoutId` to table {{%globalsets}} ... done (time: 0.053s)
    > create index idx_uukjsrfvjendwopopqryuiwmrdvwejkdhbfs on {{%globalsets}} (sortOrder) ... done (time: 0.026s)
Exception: Undefined index: name (/[...]/vendor/craftcms/cms/src/services/Globals.php:335)
#0 /[...]/vendor/craftcms/cms/src/services/Globals.php(335): yii\base\ErrorHandler->handleError(8, 'Undefined index...', '/Users/[...]...', 335, Array)
#1 /[...]/vendor/craftcms/cms/src/base/ApplicationTrait.php(1668): craft\services\Globals->handleChangedGlobalSet(Object(craft\events\ConfigEvent))
#2 /[...]/vendor/craftcms/cms/src/services/ProjectConfig.php(1189): craft\console\Application->craft\base\{closure}(Object(craft\events\ConfigEvent))
#3 [internal function]: craft\services\ProjectConfig->handleChangeEvent(Object(craft\events\ConfigEvent))
#4 /[...]/vendor/yiisoft/yii2/base/Component.php(628): call_user_func(Array, Object(craft\events\ConfigEvent))
#5 /[...]/vendor/craftcms/cms/src/services/ProjectConfig.php(752): yii\base\Component->trigger('updateItem', Object(craft\events\ConfigEvent))
#6 /[...]/vendor/craftcms/cms/src/services/ProjectConfig.php(1179): craft\services\ProjectConfig->_processConfigChangesInternal('globalSets.639d...')
#7 [internal function]: craft\services\ProjectConfig->handleChangeEvent(Object(craft\events\ConfigEvent))
#8 /[...]/vendor/yiisoft/yii2/base/Component.php(628): call_user_func(Array, Object(craft\events\ConfigEvent))
#9 /[...]/vendor/craftcms/cms/src/services/ProjectConfig.php(749): yii\base\Component->trigger('addItem', Object(craft\events\ConfigEvent))
#10 /[...]/vendor/craftcms/cms/src/services/ProjectConfig.php(521): craft\services\ProjectConfig->_processConfigChangesInternal('globalSets.639d...', true, NULL)
#11 /[...]/vendor/craftcms/cms/src/migrations/m210613_145522_sortable_global_sets.php(38): craft\services\ProjectConfig->set('globalSets.639d...', 2)
#12 /[...]/vendor/craftcms/cms/src/db/Migration.php(52): craft\migrations\m210613_145522_sortable_global_sets->safeUp()
#13 /[...]/vendor/yiisoft/yii2/console/controllers/BaseMigrateController.php(744): craft\db\Migration->up()
#14 /[...]/vendor/craftcms/cms/src/console/controllers/MigrateController.php(379): yii\console\controllers\BaseMigrateController->migrateUp('m210613_145522_...')
#15 [internal function]: craft\console\controllers\MigrateController->actionAll()
#16 /[...]/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#17 /[...]/vendor/yiisoft/yii2/base/Controller.php(181): yii\base\InlineAction->runWithParams(Array)
#18 /[...]/vendor/yiisoft/yii2/console/Controller.php(184): yii\base\Controller->runAction('all', Array)
#19 /[...]/vendor/yiisoft/yii2/base/Module.php(534): yii\console\Controller->runAction('all', Array)
#20 /[...]/vendor/yiisoft/yii2/console/Application.php(181): yii\base\Module->runAction('migrate/all', Array)
#21 /[...]/vendor/craftcms/cms/src/console/Application.php(89): yii\console\Application->runAction('migrate/all', Array)
#22 /[...]/vendor/yiisoft/yii2/console/Application.php(148): craft\console\Application->runAction('migrate/all', Array)
#23 /[...]/vendor/yiisoft/yii2/base/Application.php(392): yii\console\Application->handleRequest(Object(craft\console\Request))
#24 /[...]/craft(27): yii\base\Application->run()
#25 {main}
*** failed to apply m210613_145522_sortable_global_sets (time: 0.091s)

```

### Steps to reproduce

1. composer update 
2. ./craft db migrate/all

### Additional info

- Craft version: 3.6.17
- PHP version: 7.4

**Comment by brandonkelly** [comment]
Not able to reproduce this. Can you send your 3.6.17 database backup and `composer.json` + `composer.lock` files into support@craftcms.com so we can look into it from there?

**Comment by croxton** [comment]
Tried again but began by clearing all soft deleted elements with `./craft gc`. But hit the same issue.

OK I have established that the reason that migration fails is because I have Project Config disabled like this `/config/app.php`:
https://craftcms.com/docs/3.x/project-config.html#manual-yaml-file-generation

If I run `php craft project-config/write` to manually update the yaml before upgrading Craft, all migrations succeed. However, I cannot reliably run that command in the production environment for this project as that has multiple instances of Craft with ephemeral filesystems, so the yaml would get generated on one of the instances but not the others.  And I'd really rather not commit the project config yaml to the repo.

**Comment by elfacht** [comment]
@croxton I'm using the `project.yml` in my project out of the box, deleted the soft-deleted elements as well, still no luck.

**Comment by croxton** [comment]
@elfacht if you regenerate your project config yaml I suspect it will succeed. However, I'm not sure that migrations should rely on a yaml file being present - it should query the db for the config since that represents the application state that is being updated.

**Comment by elfacht** [comment]
@elfacht Ok, thanks! But I guess doing this on production would be a bit wild.

**Comment by brandonkelly** [comment]
@croxton It shouldn’t have been necessary to run that command first. Would still appreciate those backup + Composer files so we can look into it.

**Comment by croxton** [comment]
@brandonkelly no problem at all I will send them over - thank you for looking at this!

**Comment by croxton** [comment]
@brandonkelly the presence of old (out-of-sync) project config yaml files causes the `m210613_145522_sortable_global_sets.php` migration to fail. If I remove the '/config/project` directory entirely then the migration succeeds. In my case this means my production environment should upgrade successfully since deployments do not contain `/config/project`.

**Comment by brandonkelly** [comment]
Sounds like we got this sorted via d4d7e089057765ac070ebd7a5572a5dffbf20a26. Thanks for your help!

**Comment by brandonkelly** [comment]
We’ve spent some time looking into this, and while we can reproduce, it’s unfortunately not something that can be easily fixed without a major ProjectConfig service refactor. So we are going to look into that for Craft 4.

In the meantime, I’ve added a warning to the [Manual YAML File Generation](https://craftcms.com/docs/3.x/project-config.html#manual-yaml-file-generation) section of the Project Config page in the docs about this.

**Comment by brandonkelly** [comment]
This has been resolved for Craft 4 (see #9660) 🎉
